### PR TITLE
add boolean, date, timestamp & binary partition types

### LINKED ISSUE #1170: feat(rust): support additional types for partition values
# Description

We currently only support integers and strings for partition values. We should support the remaining types described at:

https://github.com/delta-io/delta/blob/master/PROTOCOL.md#partition-value-serialization

This is currently implemented in this function: https://github.com/delta-io/delta-rs/blob/fac6566a2c7f167db68a3723db20c54974164281/rust/src/writer/utils.rs#L122-L152

**Use Case**

**Related Issue(s)**

### PR DESCRIPTION
**Pull request by marijncv**
# Description
Adds boolean, date, timestamp & binary partition value types

# Related Issue(s)
closes #1170

### CODE REVIEW

File: rust/src/writer/utils.rs [hunk 1]
```diff
@@ -235,3 +282,113 @@ impl Write for ShareableBuffer {
...
+        BinaryArray, BooleanArray, Date32Array, Date64Array, Int16Array, Int32Array, Int64Array,
+        Int8Array, LargeBinaryArray, StringArray, TimestampMicrosecondArray,
+        TimestampMillisecondArray, TimestampNanosecondArray, TimestampSecondArray, UInt16Array,
+        UInt32Array, UInt64Array, UInt8Array,
+    };
+
+    #[test]
+    fn test_stringified_partition_value() {
+        let reference_pairs: Vec<(Arc<dyn Array>, Option<String>)> = vec![
+            (Arc::new(Int8Array::from(vec![None, Some(2)])), None),
```
Comment (wjones127) [inline rust/src/writer/utils.rs:299]: I think we technically don't care to test the behavior that we only use the first value; it's possible a future version of this function will error if not all values are identical. Could you change each case to just have a array 1 value long?
```suggestion
            (Arc::new(Int8Array::from(vec![None])), None),
```

**Review by wjones127 — CHANGES_REQUESTED** [review]
Thanks for working on this @marijncv!

Just one comment, but otherwise looks great.

**Review by fvaleye — APPROVED** [review]
LGTM

File: rust/src/writer/utils.rs [hunk 2]
```diff
@@ -235,3 +282,113 @@ impl Write for ShareableBuffer {
...
+                Some(String::from("1")),
+            ),
+            (
+                Arc::new(Int16Array::from(vec![Some(1), Some(2)])),
+                Some(String::from("1")),
+            ),
+            (
+                Arc::new(Int32Array::from(vec![Some(1), Some(2)])),
+                Some(String::from("1")),
+            ),
```
Comment (wjones127) [inline rust/src/writer/utils.rs:311]:
```suggestion
            (Arc::new(Int8Array::from(vec![1])), Some("1")),
            (Arc::new(Int16Array::from(vec![1])), Some("1")),
            (Arc::new(Int32Array::from(vec![1])), Some("1")),
```

File: rust/src/writer/utils.rs [hunk 3]
```diff
@@ -235,3 +282,113 @@ impl Write for ShareableBuffer {
...
+    use arrow::array::{
+        BinaryArray, BooleanArray, Date32Array, Date64Array, Int16Array, Int32Array, Int64Array,
+        Int8Array, LargeBinaryArray, StringArray, TimestampMicrosecondArray,
+        TimestampMillisecondArray, TimestampNanosecondArray, TimestampSecondArray, UInt16Array,
+        UInt32Array, UInt64Array, UInt8Array,
+    };
+
+    #[test]
+    fn test_stringified_partition_value() {
+        let reference_pairs: Vec<(Arc<dyn Array>, Option<String>)> = vec![
```
Comment (wjones127) [inline rust/src/writer/utils.rs:298]:
```suggestion
        let reference_pairs: Vec<(Arc<dyn Array>, Option<&str>)> = vec![
```

File: rust/src/writer/utils.rs [hunk 4]
```diff
@@ -235,3 +282,113 @@ impl Write for ShareableBuffer {
...
+                Arc::new(LargeBinaryArray::from_vec(vec![b"1", b"2"])),
+                Some(String::from("1")),
+            ),
+            (
+                Arc::new(LargeBinaryArray::from_vec(vec![b"\x00\\", b"2"])),
+                Some(String::from("\\x00\\\\")),
+            ),
+        ];
+        for (vals, result) in reference_pairs {
+            assert_eq!(stringified_partition_value(&vals).unwrap(), result)
```
Comment (wjones127) [inline rust/src/writer/utils.rs:391]: You can use `as_deref()` to turn `Option<String>` into `Option<&str>`:
```suggestion
            assert_eq!(
                stringified_partition_value(&vals).unwrap().as_deref(),
                result
            )
```

File: rust/src/writer/utils.rs [hunk 5]
```diff
@@ -235,3 +282,113 @@ impl Write for ShareableBuffer {
...
+                Arc::new(UInt64Array::from(vec![Some(1), Some(2)])),
+                Some(String::from("1")),
+            ),
+            (
+                Arc::new(StringArray::from(vec![
+                    Some(String::from("1")),
+                    Some(String::from("2")),
+                ])),
+                Some(String::from("1")),
+            ),
```
Comment (wjones127) [inline rust/src/writer/utils.rs:338]: The `StringArray::from` implementation can take a `&str`:

```suggestion
            (Arc::new(StringArray::from(vec!["1"])), Some("1")),
```

**Review by wjones127 — CHANGES_REQUESTED** [review]
Sorry for the delay. I meant my earlier comment about not needing more than 1 value in the test to apply to all the test cases, not just the one I had made a suggestion on.

Added a few other comments to make the test cases a little more compact.

**Review by roeap — APPROVED** [review]
LGTM!

**Review by wjones127 — APPROVED** [review]
Great! Thanks @marijncv

**Pull request merged**